core/src/avm2.rs

```diff
@@ -218,7 +221,8 @@ impl<'gc> Avm2<'gc> {
         Self {
             player_version,
             player_runtime,
-            stack: Vec::new(),
+            stack: vec![Value::Undefined; PREALLOCATED_STACK_SIZE],
```

Comment:
I think we can make the stack a wrapper type around `Vec<Value>` that just unsafely implements `Collect` with the mentioned assertion.

Reply:
Oh yeah, that's a simpler idea too :)

Reply:
Actually this has another small advantage, we can move all the stack-related methods to `.avm2.stack.push(5)` instead of having all of them in toplevel `avm2.push()`
(and who knows, this might be somehow a small step towards having interpreter loop have closer access to stack altogether :) )

Reply:
> I think we can make the stack a wrapper type around `Vec<Value>` that just unsafely implements `Collect` with the mentioned assertion.

I don't even think this need any wrapper at all: the stack will be always empty when GC runs, so tracing it should always be a noop.

Reply:
@moulins the `Vec` will not be empty; it will hold leftover values and `Undefined` (since we never clear it). We need either a custom `Collect` impl or adding the `#[unsafe_skip_trace]` attribute that Adrian mentioned.

Reply:
I think you misunderstand what I'm suggesting: instead of having `stack: Vec<_>, stack_i: usize`, just have a single `Vec<_>`, and manipulate it "as usual", so that its length is always the current top of the stack. (and to avoid reallocation when pushing values, we use (a reimplementation of) `Vec::push_within_capacity`)

Reply:
Yeah, this looks like it could have merit :D 
In my godbolt experiment, push with `push_with_capacity` appears to generate comparable, if not very slightly better (might be an artifact of the experiment), code: https://godbolt.org/z/ozPer6s6q

Reply:
There's also another difference - our current `pop` doesn't check for stack underflows at all, while `stack.pop()` will. I'm assuming we do want it checked, so re-adding the missing check will revert at last some % of the perf win.